### PR TITLE
chore: add K8s event recorder and events for nim operator CRDs

### PR DESCRIPTION
**Pull request by slu2011**

### CODE REVIEW

File: internal/controller/nimpipeline_controller.go [hunk 1]
```diff
@@ -339,6 +341,9 @@ func (r *NIMPipelineReconciler) updateStatus(ctx context.Context, nimPipeline *a
 	nimPipeline.Status.State = overallState
 	nimPipeline.Status.States = serviceStates
 
+	r.GetEventRecorder().Eventf(nimPipeline, corev1.EventTypeNormal, overallState,
+		"nimPipeline %s in namespace %s status %s, service states %v", nimPipeline.Name, nimPipeline.Namespace, overallState, serviceStates)
```
Comment (mikemckiernan) [inline internal/controller/nimpipeline_controller.go:345]: total nit: `NIMPipeline` would match the error on line 348.

File: internal/controller/nimcache_controller.go [hunk 2]
```diff
@@ -82,6 +84,7 @@ type NIMCacheReconciler struct {
 	log      logr.Logger
 	Platform platform.Platform
 	updater  conditions.Updater
+	recorder record.EventRecorder
```
Comment (shivamerla) [inline internal/controller/nimcache_controller.go:87]: Where is this initialized? When we setup reconciler we need to initialize this with `recorder: mgr.GetEventRecorderFor("NIMCache"),

**Comment by shivamerla** [comment]
@slu2011 RBAC needs to be updated too

```
E0926 15:56:12.656001       1 event.go:359] "Server rejected event (will not retry!)" err="events \"my-nim-pipeline.17f8d64691e223fb\" is forbidden: User \"system:serviceaccount:nim-operator:k8s-nim-operator\" cannot patch resource \"events\" in API group \"\" in the namespace \"nim-service\"" event="&Event{ObjectMeta:{my-nim-pipeline.17f8d64691e223fb  nim-service    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},InvolvedObject:ObjectReference{Kind:NIMPipeline,Namespace:nim-service,Name:my-nim-pipeline,UID:1894b4d8-c857-4387-807a-5c6d24f7a5e9,APIVersion:apps.nvidia.com/v1alpha1,ResourceVersion:1582400,FieldPath:,},Reason:Ready,Message:nimPipeline my-nim-pipeline in namespace nim-service status Ready, service states map[meta-llama3-8b-instruct-h100:Ready],Source:EventSource{Component:nimpipeline-controller,Host:,},FirstTimestamp:2024-09-26 15:54:15.680156667 +0000 UTC m=+25.904429536,LastTimestamp:2024-09-26 15:56:12.653294668 +0000 UTC m=+142.877567537,Count:4,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:nimpipeline-controller,ReportingInstance:,}"
```

File: internal/controller/nimcache_controller.go [hunk 3]
```diff
@@ -119,31 +122,43 @@ func NewNIMCacheReconciler(client client.Client, scheme *runtime.Scheme, log log
...
 	logger.Info("Reconciling", "NIMCache", nimCache.Name)
+	previousStatusState := nimCache.Status.State
 
+	defer func() {
+		if err != nil {
+			r.GetEventRecorder().Eventf(nimCache, corev1.EventTypeWarning, appsv1alpha1.NimCacheConditionReconcileFailed,
+				"NIMCache %s in namespace %s reconcile failed, msg: %s", nimCache.Name, nimCache.Namespace, err.Error())
+		} else if previousStatusState != nimCache.Status.State {
+			r.GetEventRecorder().Eventf(nimCache, corev1.EventTypeNormal, nimCache.Status.State,
+				"NIMCache %s in namespace %s reconcile success, new state: %s", nimCache.Name, nimCache.Namespace, nimCache.Status.State)
```
Comment (shivamerla) [inline internal/controller/nimcache_controller.go:145]: we can omit logging namespace as these are all namespace scoped events. Events already have a `Namespace` field associated with them.

Comment (shivamerla) [reply internal/controller/nimcache_controller.go:145]:
```$ kubectl  get events -A
NAMESPACE      LAST SEEN   TYPE      REASON              OBJECT                                                MESSAGE
nim-service    14m         Normal    Ready               nimpipeline/my-nim-pipelin1                           nimPipeline my-nim-pipelin1 in namespace nim-service status Ready, service states map[embedding:Ready reranking:Ready]
nim-service    14m         Normal    Ready               nimservice/reranking                                  NIMService reranking in namespace nim-service ready, msg: deployment "reranking" successfully rolled out

File: internal/controller/nimcache_controller.go (hunk 2 above)
Comment (slu2011) [reply internal/controller/nimcache_controller.go:87]: Exactly. See for example https://github.com/NVIDIA/k8s-nim-operator/pull/155/files/cc926309989b6822baff489cd3cfdc91f4eb3697#diff-cf3e8a734e3ad6a908183ede7ac138abd176264cd8be708b7eedceed8b387ba0R193

**Review by shivamerla — APPROVED** [review]
lgtm

File: deployments/helm/k8s-nim-operator/values.yaml [hunk 4]
```diff
@@ -7,7 +7,7 @@ operator:
     pullPolicy: Always
   args:
     - --health-probe-bind-address=:8081
-    - --metrics-bind-address=127.0.0.1:8080
+    - --metrics-bind-address=0.0.0.0:8080
```
Comment (shivamerla) [inline deployments/helm/k8s-nim-operator/values.yaml:10]: as discussed, please revert these

**Review by shivamerla — APPROVED** [review]

**Pull request merged**